### PR TITLE
Support one-time json parsing

### PR DESCRIPTION
**Pull request by gheine**

### CODE REVIEW

**Comment by jlogeart** [comment]
If I understand the diff correctly, the main change is `transform` now returns a `Seq[V]` and `processBatch` takes a `Seq[V]`.

Even though it is an improvement vs existing, I am not sure we are getting a lot of mileage as I don't think `transform` is used often. I am also not convinced it should be used as a way of parsing `Record` vs having a dedicated function that does this instead, like `read(r: Record): V` (or some implicit). And `transform` remains a function allowing transformations, filters, etc.

WDYT?

**Comment by gheine** [comment]
The problem is  regardless of whether transform is used or not, in fact I'm not sure wer're using transform anywhere right now. The problem is that we regularly need to parse record.js twice, generally in accepts and in processBatch. This way allows you to do that transformation (hence "transform") only once. Would calling it `collect` sound better?

File: app/io/flow/event/v2/actors/TypedPollActor.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  * To extend this class:
+  *   - implement system, queue, process(record)
+  *   - call start(...) w/ the name of the execution context to use
+  */
+trait TypedPollActor[V] extends TypedPollActorBatch[V] {
+  override def transform(records: Seq[Record]): Seq[V] = {
+    records.flatMap(transform)
+  }
+
+  def transform(record: Record): Option[V]
```
Comment (jackl) [inline app/io/flow/event/v2/actors/TypedPollActor.scala:20]: I'd expect transform to return a V, perhaps call it filter?

File: app/io/flow/event/v2/actors/TypedPollActorBatch.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  /**
+   * Called once for every batch read off the stream
+   */
+  def processBatch(records: Seq[V]): Unit
+
+  /**
+   * Allows the retrieved Seq[Record] to be transformed (filtered, modified, ...) before being processed
+   * This function can also be used to filter records
+   */
+  def transform(records: Seq[Record]): Seq[V]
```
Comment (jackl) [inline app/io/flow/event/v2/actors/TypedPollActorBatch.scala:30]: Yes the naming is hard - would 'map' or 'extract'  be any better in that it extracts a V from a record or maps a record to a V?

File: app/io/flow/event/v2/actors/TypedPollActorBatch.scala [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+  def system: ActorSystem
+
+  def queue: Queue
+
+  def logger: RollbarLogger
+
+  private implicit lazy val configuredRollbar: RollbarLogger = logger.fingerprint("PollActorBatch").withKeyValue("class", getClass.getName)
```
Comment (jackl) [inline app/io/flow/event/v2/actors/TypedPollActorBatch.scala:38]: TypedPollActorBatch

**Comment by jlogeart** [comment]
How about leveraging `Reads[T]` instead of leaving it manual to the concrete class?

```scala
trait TypedPollActorBatch[V] extends Actor with ActorLogging {
  implicit val reads: Reads[V]
  // ...

    val transformedRecords = records.flatMap(_.js.asOpt[V](reads))
    if (transformedRecords.nonEmpty) {
      processBatch(transformedRecords)
    }
```

So that all that's necessary is to provide an implicit Reads[V] in the implementation.

File: app/io/flow/event/v2/actors/TypedPollActor.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+/**
+  * [[TypedPollActor]] periodically polls a kinesis stream (by default every 5
+  * seconds), invoking process once per message.
+  *
+  * To extend this class:
+  *   - implement system, queue, process(record)
+  *   - call start(...) w/ the name of the execution context to use
+  */
+trait TypedPollActor[V] extends TypedPollActorBatch[V] {
+  override def transform(records: Seq[Record]): Seq[V] = {
```
[the remark this answers is not in the document]
Comment (gheine) [reply app/io/flow/event/v2/actors/TypedPollActor.scala:16]: Agree that would be a better name, but it's called `transform` now, and not sure how many times it is overridden, so would prefer to keep the same.

File: app/io/flow/event/v2/actors/TypedPollActor.scala (hunk 1 above)
Comment (gheine) [reply app/io/flow/event/v2/actors/TypedPollActor.scala:20]: filter functions usually return a boolean 🤷

**Comment by gheine** [comment]
I like it. That would be new subtypes of TypedPollActorBatch so that we can still support the current PollActorBatch, where `transform` returns the Record type.

**Pull request merged**